base/iobuffer.jl

```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
     return io
 end
 
+# TODO: Should check for seekable and error if not
 function seekend(io::GenericIOBuffer)
     io.ptr = io.size+1
     return io
 end
 
-# choose a resize strategy based on whether `resize!` is defined:
-# for a Vector, we use `resize!`, but for most other types,
-# this calls `similar`+copy
-function _resize!(io::GenericIOBuffer, sz::Int)
-    a = io.data
-    offset = io.offset
-    if applicable(resize!, a, sz)
-        if offset != 0
-            size = io.size
-            size > offset && copyto!(a, 1, a, offset + 1, min(sz, size - offset))
-            io.ptr -= offset
-            io.size -= offset
-            io.offset = 0
-        end
-        resize!(a, sz)
+# Resize data to exactly size `sz`. Either resize the underlying data,
+# or allocate a new one and copy.
+# This should only be called after the offset is zero - any operation which calls
+# _resize! should reset offset before so.
+function _resize!(io::GenericIOBuffer, new_size::Int)
+    old_data = io.data
+    if applicable(resize!, old_data, new_size)
+        resize!(old_data, new_size)
```

Comment:
This is already defined to call `overallocation`, so now we're calling it twice (doubling the wasted space) if anyone was using an IOBuffer wrapping something mutable. Could you sink the call to `overallocation` into this function, and instead add another (exact::Bool) that controls whether it gets called? That may also avoid needing to add the requirement to call `offset = 0` first, which can result in redundant and unnecessary data copies in some cases in the old buffer, when the data needed to be moved to the new buffer anyways.